src/java/org/apache/cassandra/service/accord/AccordJournal.java

```diff
@@ -0,0 +1,407 @@
+ * limitations under the License.
+ */
+package org.apache.cassandra.service.accord;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.*;
+import java.util.concurrent.Executor;
+import java.util.zip.Checksum;
+
+import com.google.common.primitives.Ints;
+
+import accord.local.Node.Id;
+import accord.local.PreLoadContext;
+import accord.messages.Accept;
+import accord.messages.Apply;
+import accord.messages.Commit;
+import accord.messages.MessageType;
+import accord.messages.PreAccept;
+import accord.messages.TxnRequest;
+import accord.primitives.*;
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.io.IVersionedSerializer;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.journal.Journal;
+import org.apache.cassandra.journal.KeySupport;
+import org.apache.cassandra.journal.Params;
+import org.apache.cassandra.journal.ValueSerializer;
+import org.apache.cassandra.service.accord.serializers.AcceptSerializers;
+import org.apache.cassandra.service.accord.serializers.ApplySerializers;
+import org.apache.cassandra.service.accord.serializers.CommitSerializers;
+import org.apache.cassandra.service.accord.serializers.EnumSerializer;
+import org.apache.cassandra.service.accord.serializers.PreacceptSerializers;
+
+import static org.apache.cassandra.db.TypeSizes.BYTE_SIZE;
+import static org.apache.cassandra.db.TypeSizes.INT_SIZE;
+import static org.apache.cassandra.db.TypeSizes.LONG_SIZE;
+import static org.apache.cassandra.utils.FBUtilities.updateChecksumInt;
+import static org.apache.cassandra.utils.FBUtilities.updateChecksumLong;
+
+/*
+ *  TODO: expose more journal params via Config
+ */
+class AccordJournal
+{
+    private static final Set<Integer> SENTINEL_HOSTS = Collections.singleton(0);
+
+    final File directory;
+    final Journal<Key, TxnRequest<?>> journal;
+
+    AccordJournal()
+    {
+        directory = new File(DatabaseDescriptor.getAccordJournalDirectory());
+        journal = new Journal<>("AccordJournal", directory, Params.DEFAULT, Key.SUPPORT, MESSAGE_SERIALIZER);
+    }
+
+    AccordJournal start()
+    {
+        journal.start();
+        return this;
+    }
+
+    void shutdown()
+    {
+        journal.shutdown();
+    }
+
+    boolean mustAppend(PreLoadContext context)
+    {
+        return context instanceof TxnRequest && Type.mustAppend((TxnRequest<?>) context);
+    }
+
+    void append(int storeId, PreLoadContext context, Executor executor, Runnable onDurable)
+    {
+        append(storeId, (TxnRequest<?>) context, executor, onDurable);
+    }
+
+    void append(int storeId, TxnRequest<?> message, Executor executor, Runnable onDurable)
+    {
+        Key key = new Key(message.txnId, Type.fromMsgType(message.type()), storeId);
+        journal.asyncWrite(key, message, SENTINEL_HOSTS, executor, onDurable);
```

Comment:
`org.apache.cassandra.journal.Metadata` whole existence is to track hosts, but our requests don't contain a host and we currently use a fixed host at the write side; what is this host tracking supposed to be for?

Reply:
Segment target prioritisation for `Compactor` (stripped from the patch for now but will be reintroduced shortly).

Reply:
where are we getting the hosts from?  If we get from `Message` then its a single host, but this API is a collection of hosts, so implies there is a source with multiple hosts?

Reply:
"Hosts" is arguably slightly too specific terminology at this point and should be renamed. It was first named when entries in all journals were expected to be tagged by owning multiple hosts.

In the case of commitlog/hint and batchlog replacement - the replicas that needed to acknowledge having applied a mutation. Once invalidated by all (or possibly a majority), the entry would be eligible to be compacted away.

Accord journal was initially supposed to be keyed by hosts as well, or at least such was my (mis?)-understanding. It looks like instead, in a follow-up patch, we'll be writing a message once per node, not per `CommandStore`, and ownership set will contain on `CommandStore` ids, not hosts.

So in the longer run I'll rename "hosts" to "owners", but perhaps later, when that follow-up happens?

Reply:
> the replicas that needed to acknowledge having applied a mutation.

Is this the replicas Accord selected, or the full set, aka w/e is in Topology?  If latter, then this is acting as a cache of what is stored in the epoch?

Reply:
Resolving, cheers.

Reply:
> Accord journal was initially supposed to be keyed by hosts as well, or at least such was my (mis?)-understanding

This may yet happen for progress log purposes